Change Build System to DMBS

NicoHood opened the pull request:
I am currently working on an improved makefile.
I therefore cleaned up the current makefile a bit bit also included DMBS as build environment.
Once all points are discussed we can merge the PR, but for now keep it please open to let others test it first ;)

### Changes to the old makefile
* Removed .a archive
* Fixed LDFLAGS
* Added LTO support
* Reordered a few things
* Removed unimportant/unused stuff

### Improvements
* Better makefile overview with DMBS
* Simpler to integrate with DMBS
* Latest avr-gcc 5.3 with LTO
* **Improves flash usage with 1.5kb (old MP firmware)**
* Bumped version number to 1.1

### Things to discuss
~~* Why was `-fno-tree-scev-cprop` used? I see no real difference~~
* Someone has to test the hex output
* ~~Do we want to fully remove the old makefile as the new DMBS makefile supersedes the current one?~~
~~* I removed the version number in the new makefile. Do we want to keep this mechanism?~~
~~* In `defines.h` you can set the MP device (normal/mini). I'd move this to the makefile if that is okay. (requires a few wrappers)~~

### Links
* https://github.com/NicoHood/AVR-Development-Environment-Script
* https://github.com/abcminiuser/dmbs

```
AVR Memory Usage (MP)
----------------
Device: atmega32u4

Program:   27166 bytes (82.9% Full)
(.text + .data + .bootloader)

Data:       1656 bytes (64.7% Full)
(.data + .bss + .noinit)


AVR Memory Usage (MP Mini)
----------------
Device: atmega32u4

Program:   22264 bytes (67.9% Full)
(.text + .data + .bootloader)

Data:       1794 bytes (70.1% Full)
(.data + .bss + .noinit)
```

limpkin commented:
- don't know about -fno-tree-scev-cprop
- version number: well this can be discussed, i'm used to use the defines.h method, but some people prefer fetching the version from github directly
- please keep the type of device set in defines.h

NicoHood commented:
Okay. So I will remove the optimization, as there is no size benefit and the only information I could find is this:
https://tty1.net/blog/2008/avr-gcc-optimisations_en.html

I think the version number is okay like this. The windows version doesnt have this anyways I think.
Definitions inside defined.h are okay for me.

I will push another commit and then finally rename the title, so you can merge if you want to. The current makefile doesnt build anyways and needs lots of improvements, so actually there is no reason why not merging ;)

limpkin commented:
@jim17 can you double check this PR?

jim17 commented:
I will check tomorrow if I have some spare time :)
El 15/4/2016 8:24 p. m., "Mathieu" <notifications@github.com> escribió:

> @jim17 <https://github.com/jim17> can you double check this PR?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/limpkin/mooltipass/pull/186#issuecomment-210576146>
>

jim17 commented:
It can be removed but, this specific flag means to don't propagate copies of tree scalar evolutions.
https://gcc.gnu.org/onlinedocs/gccint/Scalar-evolutions.html

NicoHood commented:
It did not made the program any smaller. However thanks for the link, I could not find any information about this.

So can the PR be merged?

limpkin commented:
@jim17 this PR is ok for you?

jim17 commented:
I really don't like the fact of having external repository dependencies (submodules). I like more the fact of having all the project needs inside the project. What happens if the submodules added are not mantained anymore or are deleted ?

The makefile you are trying to commit is more clean, but I hink the old one has all the needed in. It does not depend on extra makefiles and making a change (atleast for me) is easier.

If the Pull Request is accepted, don't forget to add a comment about how to update and download the submodules you have added to project.

Also as a note you're currently placing the compiler inside AVR-Development-Environment-Script folder, so what happens if someone have the compiler in a different location ?

NicoHood commented:
You can alwazs ship a copy of the dmbs if you like to.
The current makefile is quite unordered and hard to maintain. I could not get it running anzmore, so I removed it completely. If you like to, you can keep it, but it will not work correct. Limpkin has to decide if he wants a submodule or rather a full clone inside the repository.

The compiler is in the special folder, because it is also a submodule. It will alwazs be downloaded in this folder. And if not, then a linux user will have the skills to see that the path is different. The new compiler saves about 1,5kb of flash. It can be used optional though. Maybe further debian packages ship avrgcc 5.x anywazs.

I suggest to add the dmbs as non submodule and make the avrgcc 5.3 optional in the makefile. This way we can comment it out and place a small instruction in the readme how to download the submodule, as it is reallz optional here.

limpkin commented:
Hello all,

You're both making very good points.
Nicohood, I'm guessing you couldn't compile just because a few files were
not included in the makefile?
It seems indeed a bit foolish to completely remove one make file for
another. Would it therefore be possible to make 2 separate folders for each
compiling environment?
Nicohood: I'll need the documentation that Jim requested.

Thanks to you both!
On Apr 20, 2016 09:06, "Nico" <notifications@github.com> wrote:

> You can alwazs ship a copy of the dmbs if you like to.
> The current makefile is quite unordered and hard to maintain. I could not
> get it running anzmore, so I removed it completely. If you like to, you can
> keep it, but it will not work correct. Limpkin has to decide if he wants a
> submodule or rather a full clone inside the repository.
>
> The compiler is in the special folder, because it is also a submodule. It
> will alwazs be downloaded in this folder. And if not, then a linux user
> will have the skills to see that the path is different. The new compiler
> saves about 1,5kb of flash. It can be used optional though. Maybe further
> debian packages ship avrgcc 5.x anywazs.
>
> I suggest to add the dmbs as non submodule and make the avrgcc 5.3
> optional in the makefile. This way we can comment it out and place a small
> instruction in the readme how to download the submodule, as it is reallz
> optional here.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/limpkin/mooltipass/pull/186#issuecomment-212289547>
>

limpkin commented:
Hey All,

Nicohood, are you ok with my above suggestion?

Thanks

NicoHood commented:
Yes,
just had not time to update it/answer.

I think all depends on: do you really want to keep the legacy makefile? They can coexist, but I am unable to fix the old, as its very complex and a mess. I can revert the old makefile, but it will be not usable. Also there is less reason for using the old one.

If you make dmbs the only makefile it would really make sense to hardcopy it (and don't submodule). You are totally right with this. But this depends on the decision above.

Documentation is not required on a hardcopy. For the avr-gcc 6.1 I will add an optional switch + explanation in the makefile.

-> So everything depends on: Do you want dmbs the new build system for linux or is it just a second option and we have to maintain two makefiles? DMBS would improve a lot of things and make them simpler (uploading, different compilers, changes, readability, output file structure, builtin docs, etc).

limpkin commented:
> I think all depends on: do you really want to keep the legacy makefile? 
We do.

>They can coexist
Great!

>  I am unable to fix the old, as its very complex and a mess.
Could you let us know exactly what's the current problem? Is it just because it is including bootloader_main.c? Miguel, could you have a look?

> Also there is less reason for using the old one.
There are the reasons mentioned above.

> is it just a second option and we have to maintain two makefiles?
yep, we'd like to keep both.

> DMBS would improve a lot of things and make them simpler (uploading, different compilers, changes, readability, output file structure, builtin docs, etc).
We're not asking you to drop DMBS...

NicoHood commented:
Great, thats a word! I will have a look at it. Give me some time though.

NicoHood commented:
I think the PR is ready to merge. I updated:
* Removed AVR Script submodule (ubuntu 16.04 has the desired avr-gcc version)
* Added Documentation for all compile options
* Readded the legacy makefile
* Bumped version number to 1.1, maybe you want even 1.2 I do not know.

Someone has to fix the old makefile still, if you wish. Dmbs works fine as it is right now and avr-gcc with lto gives a lot smaller binary!

limpkin commented on source_code/src/Readme.md:
```diff
@@ -0,0 +1,39 @@
+
+### Legacy makefile
+The legacy makefile is located in `source_code`. Run `make` to
+compile the firmware. It might be outdated, as it is harder to maintain.
```
please remove the "harder to maintain" part.

NicoHood replied:
Why? Its the truth.

limpkin replied:
it's also highly subjective. Either you remove it or I'll remove it.

limpkin commented:
Feel free to correct me, but did you actually modify our current makefile (source_code/Makefile) yet you mention that it still has to be fixed?

NicoHood commented:
I fixed the old makefile. But then you added the bootloader which broke it again. And I was unable to fix it again. So I left it as it is, half updated.

limpkin commented:
How come were you unable? what eludes you?

limpkin commented:
pinging @jim17  for double checks

NicoHood commented:
I did a quick check and it seems there are overlapsing sections. I am not sure of the reason anymore and I had my reasons to not fix it (mostly because of the makefile knowledge and since dmbs was better anyways).

So the current makefile is broken anyways. Would you please merge this PR so I can continue working on future firmware/bootloader fixes? I cannot commit fixes, as my branch with fixes will always contain the dmbs already, as I really need a build system under linux.

limpkin commented:
Then if it doesn't compile with our current makefile, please reverse the changes you made to it.

NicoHood commented:
I see no reason. I just cleaned things up and made the output a lot smaller. There might be just a tiny thing that needs to be resolved. It is still better, even though it is broken.

Actually you can remove it. I see no reason why you want to keep it. It just messed up the folders with .o and .d files. And if someone really want to look at the old file: we got git, and this PR documents the change. Its so simple to find the old makefile for reference. But there is not a single reason why to keep it.

And if @jim17 cannot fix it (or doesnt want to), I am the only one who really needs the makefile. And other (possibly new) tinkerers will use the new makefile anyways. Noone needs the old makefile to be honest. You said the community will fix it. I (the community) fixed it just a little different and you want to hold on on useless/unused old stuff. Its been 2 month now since I originally pushed a fix.

Yes I am frustrated.

limpkin commented:
@NicoHood We **will not accept pull requests that actually don't fix things**. Either you commit a change to our old makefile that allows our solution to compile, or you don't commit any change.

>> Noone needs the old makefile to be honest.
Please refrain from making broad and absolute statements like these.

NicoHood commented:
The PR fixes the old makefile partly, but not completely.
The PR add an alternative makefile which completely fixes an addresses all issues of the old one.
The old one is obsolete, thats all.

NicoHood commented:
I've updated the PR finally. Now **all** problems should be solved.
* The old makefile works fully with the same result in size as the new makefile
* The new makefile use only one submodule, which is documented
* avr-gcc with lto is available in ubuntu 16.04, so not custom script is required anymore.
